### ISSUE TITLE
Unstable Vital Dynamics

### ISSUE THREAD

**Issue by daniel-klein** [starsimhub/starsim#695]
Consider a simulation with the following demographics
```python
dems = [
    ss.Births(pars=dict(birth_rate=20)),
    ss.Deaths(pars=dict(death_rate=20))
]
```
One might expect a stable population. However, the result is a dramatically declining population.
<img width="612" alt="image" src="https://github.com/user-attachments/assets/d1c127a0-4bee-4e7f-9382-d2399fb7ee25">

This result is for two reasons:
1. `Births` takes the floor when deterministically adding agents.
`n_new = int(np.floor(sim.people.alive.count() * scaled_birth_prob))`

2. Module order matters. Because `Births` comes before `Deaths` in the list of `dems`, agents are added to the population (resulting in a larger population), and then `Deaths` applies the death rate to _all_ agents, including the new agents. A 1% increase followed by a 1% decrease results in a net decline.

Switching the order of demographic module updates (and changing the Births floor to something like `sc.randround`) resolves the issue. Why? Because `Deaths` simply schedules agents for death via `ti_death` without actually removing them from the population. The `Births` module then adds agents considering the _same denominator_ as used for deaths.
<img width="612" alt="image" src="https://github.com/user-attachments/assets/70d0173b-cae4-4377-a5db-c898b065e556">

We cannot expect users to know they have to put `Deaths` before `Births` to get a stable population.

Next Steps:
- [ ] Change `Births` away from `floor`. It could apply a probability per-agent
- [ ] Consider a combined `BirthDeath` module that would act on the same denominator.
- [ ] Consider only adding births (including from `Pregnancy`) at the end of the time step, although that will likely create other problems.
- [ ] Consider excluding agents born on the current time step from the death rate (although doing so could have other unintended consequences as we know postnatal mortality is typically quite high.)
- [ ] Consider adjusting the death rate by the birth rate, if `Deaths` follows `Births`.

Thanks to Deven for raising this issue in the context of SEIRD in TBsim.

**Comment by cliffckerr** [comment]
We had discussed this issue previously here:
https://github.com/starsimhub/starsim/issues/303

I agree with the logic that newborns have a nonzero mortality risk, so this makes sense. It's like how the population replacement rate is closer to 2.1 than 2.0 births per woman -- nonintuitive, but correct.

I'm also not sure I'd call the drop "dramatic": take a look at this:
```py
import starsim as ss
import matplotlib.pyplot as plt

dems = [
    ss.Births(pars=dict(birth_rate=20)),
    ss.Deaths(pars=dict(death_rate=20))
]

sim = ss.Sim(demographics=dems, dur=30)
sim.run()

res = sim.results
plt.plot(res.timevec, res.n_alive)
plt.ylim(bottom=0, top=10_000)
plt.show()
```
![image](https://github.com/user-attachments/assets/702437a6-88a6-44ac-a58a-4113d9d8f505)

**Comment by vgdeven** [comment]
Just curious, why not have a separation of labor between - calculation/drawing of transitions (eg. n_births, n_deaths etc.) in one step and then application of transitions  (eg., ΔState = n_births-n_deaths) in other?  That way all transitions will based on the same denominator?

**Comment by cliffckerr** [comment]
Good question! Two main reasons:
1. Since different modules can have different timesteps, there isn't a single point at which the deltas can be collected, and another point at which they can be applied. (Births and deaths could even use different timesteps, although that would probably not be a wise choice :))
2. The model logic is intended so that under "standard" conditions, things can take effect immediately. For example, on a single timestep, a person can be tested, test positive, be scheduled for treatment, receive treatment, and then be less likely to infect someone else (or die etc). If we collected all the deltas and then modified the states after, then each step would introduce a 1-step delay (e.g. if they are tested on day `t`, they couldn't start treatment until day `t+1`, and then this wouldn't impact their probability of death/infectiousness until day `t+2`).

**Comment by daniel-klein** [comment]
Agree the drop is not dramatic, but will still come as a surprise to users.

**Comment by daniel-klein** [comment]
We should Births away from the `np.floor`. It could look more like a simple version of `Pregnancy` and could even link newborns and mothers - just without bothering with the pre/post-natal periods nor vertical network edges. I suppose one could configure Pregnancy in this way, but we could just add a shim layer to help the user if not fully refactoring as BasePregnancy and Pregnancy.

**Comment by daniel-klein** [comment]
We could also change example scripts to place `Deaths` before `Births` and describe why this change matters in a tutorial on demographics.

**Comment by cliffckerr** [comment]
I had a look at this and I think something else is going on. Partly, we wouldn't expect them to match anyway, since 1.1*0.9 != 1.0. The order shouldn't matter, but does:
```py
s1 = ss.Sim(label='Births→Deaths', demographics=[ss.Births(birth_rate=20), ss.Deaths(death_rate=20)])
s2 = ss.Sim(label='Deaths→Births', demographics=[ss.Deaths(death_rate=20), ss.Births(birth_rate=20)])
msim = ss.parallel(s1, s2)
msim.plot()
```
![image](https://github.com/user-attachments/assets/1aeb4e5d-746b-4573-8f4a-e44dd4deec14)

**Comment by daniel-klein** [comment]
The floor in `Births` magnifies the difference resulting from the order. Surprising to see the difference in variance though. In my exploration, I replaced the floor with a round and that did help the drift. But I didn't look at variance.

**Comment by cliffckerr** [comment]
This is with `floor` removed, which is why I think it's something else happening.

**Comment by cliffckerr** [comment]
Closed by #749
